atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/consistency/TimestampCorroborationConsistencyCheck.java

```diff
@@ -0,0 +1,88 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.impl.consistency;
+
+import java.util.function.LongSupplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.palantir.atlasdb.factory.ImmutableTransactionManagerConsistencyResult;
+import com.palantir.atlasdb.factory.TransactionManagerConsistencyResult;
+import com.palantir.logsafe.SafeArg;
+
+public class TimestampCorroborationConsistencyCheck implements TransactionManagerConsistencyCheck {
+    private static final Logger log = LoggerFactory.getLogger(TimestampCorroborationConsistencyCheck.class);
+
+    private final LongSupplier conservativeBound;
+    private final LongSupplier freshTimestampSource;
+
+    public TimestampCorroborationConsistencyCheck(LongSupplier conservativeBound, LongSupplier freshTimestampSource) {
+        this.conservativeBound = conservativeBound;
+        this.freshTimestampSource = freshTimestampSource;
+    }
+
+    @Override
+    public TransactionManagerConsistencyResult get() {
+        // The ordering is important, because if we get a timestamp first, we may have a false positive if we have
+        // a long GC between grabbing the fresh timestamp and the lower bound (e.g. if someone punches in between).
+        long lowerBound;
+        long freshTimestamp;
+
+        try {
+            lowerBound = conservativeBound.getAsLong();
+        } catch (Exception e) {
+            log.warn("Could not obtain a lower bound on timestamps, so we don't know if our transaction manager"
+                    + "is consistent");
+            return indeterminateResultForException(e);
+        }
+
+        try {
+            freshTimestamp = freshTimestampSource.getAsLong();
+        } catch (Exception e) {
+            log.warn("Could not obtain a fresh timestamp, so we don't know if our transaction manager is"
+                    + " consistent.");
+            return indeterminateResultForException(e);
+        }
+
+        if (freshTimestamp < lowerBound) {
+            log.error("Your AtlasDB client believes that the unreadable timestamp was {}, but that's newer than a"
+                    + " fresh timestamp of {}, which implies clocks went back. If using TimeLock, this could be because"
+                    + " timestamp bounds were not migrated properly - which can happen if you've moved TimeLock"
+                    + " Server without moving its persistent state. For safety, AtlasDB will refuse to start.",
+                    SafeArg.of("unreadableTimestamp", lowerBound),
+                    SafeArg.of("freshTimestamp", freshTimestamp));
+            return ImmutableTransactionManagerConsistencyResult.builder()
```

Comment:
We're not catching an exception here, but it would be nice if we had a reason other than unknown to put in the result to make the eventual exception more legible.

Reply:
OK. I've created a 'dummy' `AssertionError` for this case that includes information about the lower bound/timestamp.